### DISCUSSION TITLE
Memory profiling of coupled model

### DISCUSSION THREAD

**Discussion by DeniseWorthen** [ufs-community/ufs-weather-model#779] [Announcements]
The cpld_control_wave test case from the develop branch of ufs-weather-model was run for 10days under the following conditions:

1. Add the code to the WW3 cap to add ESMF memory profiling (see [branch](https://github.com/DeniseWorthen/WW3/tree/feature/memprofile2))
2. Reduce the wave PEs from 200 (!) to 20
3. Turn on ProfileMemory=true for CICE, MOM6 and WW3

I created a small script to pull the VmPeak value at the end of the Model_Advance phase for each PET for ICE, OCN and WW3. The Memory use for ICE is shown below. The % increase over 240 hours for CICE ranges between 0.13 to 0.95

<img width="770" alt="ice memory 240hr" src="https://user-images.githubusercontent.com/40498404/131356460-94b2abe0-c7cd-46ff-b0b0-73ac72346e18.png">

The Memory use for W3 is shown below. The % increase over 240 hours for W3 ranges between 0.27 to 1.9

<img width="754" alt="wav memory 240hr" src="https://user-images.githubusercontent.com/40498404/131358525-dbb2a4d3-f838-4b4b-b733-0cd7e0c35670.png">

The Memory use for OCN is shown below. The % increase over 240 hours for MOM6 ranges between 0.0 and 2.6. If the final memory use value is excluded however, the maximum % increase is reduced to 1.3. This removes the final "uptick" of memory use on PE 150. (Is this is what GFDL was referring to when they said there was a memory leak associated with the Finalize?). See issue #764 

<img width="797" alt="ocn memory 240hr" src="https://user-images.githubusercontent.com/40498404/131369686-c401a102-1a3f-45d4-a79a-1ba9dfaa83de.png">

**Comment by jiandewang** [comment]
I repeated Denise's run and re-plotted the memory usage in a slight different way, we can see there is ~1% increase of memory usage in this 10day run. Note here PE150 is not included in the plot.  Bob explained that the sudden jump in Denise's previous plot in PE150 is due to writing of restart file in the model ending stage
![base](https://user-images.githubusercontent.com/14180427/133946309-f9781f8c-1c22-4dd6-b4d5-029d88ad936c.png)
e.

**Comment by jiandewang** [comment]
this figure is based on the same 10 day run but only save last timestep of ocean results, the memory usage became flat
![snapshot](https://user-images.githubusercontent.com/14180427/133946431-799d4da1-72f9-4de3-a64a-13fab6d8075a.png)
.

**Reply by DeniseWorthen** [reply]
@jiandewang Very interesting. Could you repeat this test (once w/ 6hr output, once w/ output only at the end) for the c384mx025 case ?

**Comment by jiandewang** [comment]
@DeniseWorthen this is from C384 OCN025 run with 6hr output frequency, note total OCN PE is 120 but here I only plotted every other 4 PEs 
![base-025](https://user-images.githubusercontent.com/14180427/134409401-4c16760d-48ca-4d7c-b4a3-ff17b5fce877.png)
y

**Comment by jiandewang** [comment]
this is the same above but save last step of snapshot
![snapshot-025](https://user-images.githubusercontent.com/14180427/134409978-ec548fea-aabe-4d26-9ec7-7e617e448095.png)
t

**Comment by MatthewMasarik-NOAA** [comment]
For WW3, the profiling output points to memory increase occurring in the call to wmwave, so internal to the model vs. in the cap.  Given this we've begun to track down the leaks by running valgrind over our WW3 regtest, ww3_ufs1.1, and starting to work through the output.
